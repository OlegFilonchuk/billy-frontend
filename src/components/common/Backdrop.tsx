import React, { FC, useEffect, useState } from 'react';
import { TWTransitionDuration } from '../../types';
import { TRANSITION_DURATION } from '../../constants';

type Props = {
  open: boolean;
  transitionDuration?: TWTransitionDuration;
  onClick: () => void;
};

const Backdrop: FC<Props> = ({
  open,
  transitionDuration = TRANSITION_DURATION,
  onClick,
}) => {
  const [show, setShow] = useState(false);

  useEffect(() => {
    if (open) {
      setShow(true);
    } else {
      setTimeout(() => {
        setShow(false);
      }, transitionDuration);
    }
  }, [open, transitionDuration]);

  return (
    <div
      role="presentation"
      className={`absolute left-0 right-0 bottom-0 bg-black transition-opacity ease-in-out duration-300 z-10 ${
        open ? 'opacity-50' : 'opacity-0'
      } ${show ? 'top-0' : 'top-full'}`}
      onClick={onClick}
    />
  );
};

export default Backdrop;
