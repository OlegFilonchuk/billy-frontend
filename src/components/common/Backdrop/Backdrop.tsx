import React, { FC, useEffect, useState } from 'react';
import { TWTransitionDuration } from '../../../types';
import { TRANSITION_DURATION } from '../../../constants';

export type BackdropProps = {
  open: boolean;
  transitionDuration?: TWTransitionDuration;
  onClick: () => void;
};

const Backdrop: FC<BackdropProps> = ({
  open,
  transitionDuration = TRANSITION_DURATION,
  onClick,
}) => {
  const [show, setShow] = useState(false);

  useEffect(() => {
    if (open) {
      setShow(true);
    } else {
      setTimeout(() => {
        setShow(false);
      }, transitionDuration);
    }
  }, [open, transitionDuration]);

  return (
    <div
      role="presentation"
      className={`absolute left-0 right-0 bottom-0 bg-black transition-opacity ease-in-out duration-${transitionDuration} z-10 ${
        open ? 'opacity-50' : 'opacity-0'
      } ${show ? 'top-0' : 'top-full'}`}
      onClick={onClick}
    />
  );
};

export default Backdrop;
