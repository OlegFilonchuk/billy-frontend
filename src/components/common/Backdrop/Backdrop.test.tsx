import { fireEvent, render, RenderResult } from '@testing-library/react';
import React from 'react';
import Backdrop from './Backdrop';

const handleClick = jest.fn();

describe('Backdrop', () => {
  let ui: RenderResult;
  let open: boolean;
  describe('in closed state', () => {
    beforeEach(() => {
      open = false;
      ui = render(<Backdrop onClick={handleClick} open={open} />);
    });

    it('should should render properly', () => {
      expect(ui.container).toMatchSnapshot();
    });

    it('should be transparent if the "open" prop is false', () => {
      const backdrop = ui.getByRole('presentation');
      expect(backdrop).toHaveClass('opacity-0');
    });
  });
  describe('in open state', () => {
    beforeEach(() => {
      open = true;
      ui = render(<Backdrop onClick={handleClick} open={open} />);
    });

    it('should should render properly', () => {
      expect(ui.container).toMatchSnapshot();
    });
    it('should be a half opaque if the "open" prop is true', () => {
      const backdrop = ui.getByRole('presentation');
      expect(backdrop).toHaveClass('opacity-50');
    });

    it('should trigger an onClick callback on click', () => {
      const backdrop = ui.getByRole('presentation');
      fireEvent.click(backdrop);
      expect(handleClick).toHaveBeenCalled();
    });
  });
});
