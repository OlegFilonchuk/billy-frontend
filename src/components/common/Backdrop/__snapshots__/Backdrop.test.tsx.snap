// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop in closed state should should render properly 1`] = `
<div>
  <div
    class="absolute left-0 right-0 bottom-0 bg-black transition-opacity ease-in-out duration-300 z-10 opacity-0 top-full"
    role="presentation"
  />
</div>
`;

exports[`Backdrop in open state should should render properly 1`] = `
<div>
  <div
    class="absolute left-0 right-0 bottom-0 bg-black transition-opacity ease-in-out duration-300 z-10 opacity-50 top-0"
    role="presentation"
  />
</div>
`;
